WIP [WFLY-12835] HealthChecks in subdeployments are not registered

cghislai opened the pull request:
Allow healthchecks to be discovered in ear deployments.

https://issues.redhat.com/browse/WFLY-12835

cghislai committed:
Add & fix tests for ear

cghislai commented:
Im not sure about 2 points:
- Is CDI.current the only and better way to get an instance of the BeanManager?
- What are the expectations concerning the integration of mp-config in mp-health healthchecks? This seems an undocumented feature, and there are undefined behaviours, for instance if a config property is defined in 2 wars in an ear, should that property be resolved for the healthchecks, and to which value? Should that change depending on the subdeployment in containing the healthcheck?

jamezp commented:
/ok-to-test

jmesnil commented on microprofile/health-smallrye/src/main/java/org/wildfly/extension/microprofile/health/deployment/CDIExtension.java:
```diff
@@ -24,26 +24,21 @@
-import javax.enterprise.inject.spi.BeforeShutdown;
-import javax.enterprise.inject.spi.Extension;
-import javax.enterprise.inject.spi.ProcessAnnotatedType;
+import javax.enterprise.inject.spi.*;
```
please do not use start imports

jmesnil commented on microprofile/health-smallrye/src/main/java/org/wildfly/extension/microprofile/health/deployment/CDIExtension.java:
```diff
@@ -98,8 +96,23 @@ private void afterDeploymentValidation(@Observes final AfterDeploymentValidation
     }
 
+    private BeanManager getBeanManager(Module module, BeanManager eventBeanManager) {
+        return tryGetBeanManagerFromJndiLookup()
```
I don't think it is necessary in the DUP to use JNDI to look up the BeanManager. Using directly `CDI.current().getBeanManager()`is fine

jmesnil commented on microprofile/health-smallrye/src/main/java/org/wildfly/extension/microprofile/health/deployment/CDIExtension.java:
```diff
@@ -74,19 +69,22 @@
+        instance = beanManager.createInstance();
+
+        // Register them to be called using this module classloader
+        addHealthChecks(HealthLiteral.INSTANCE, reporter::addHealthCheck, module, healthChecks);
```
module does not need to be passed to these methods. Current code is already taking the `module.getClassLoader()`

jmesnil commented on microprofile/health-smallrye/src/main/java/org/wildfly/extension/microprofile/health/deployment/DeploymentProcessor.java:
```diff
@@ -35,8 +33,10 @@
 import org.wildfly.extension.microprofile.health.HealthReporter;
 import org.wildfly.extension.microprofile.health.MicroProfileHealthSubsystemDefinition;
 import org.wildfly.extension.microprofile.health._private.MicroProfileHealthLogger;
+import static org.jboss.as.weld.Capabilities.WELD_CAPABILITY_NAME;
```
please do not change the order of imports

jmesnil commented on testsuite/integration/basic/src/test/java/org/wildfly/test/integration/microprofile/health/MicroProfileHealthTestBase.java:
```diff
@@ -37,10 +37,7 @@
-import org.jboss.arquillian.container.test.api.Deployment;
-import org.jboss.arquillian.container.test.api.OperateOnDeployment;
-import org.jboss.arquillian.container.test.api.RunAsClient;
+import org.jboss.arquillian.container.test.api.*;
```
please do not use star imports

jmesnil commented on testsuite/integration/basic/src/test/java/org/wildfly/test/integration/microprofile/health/MicroProfileHealthTestBase.java:
```diff
@@ -60,15 +59,37 @@
 @RunAsClient
 public abstract class MicroProfileHealthTestBase {
```
This test should not be modified, it verifies the behaviour for a WAR deployment.
Checking EAP  deployment should be done by another test

jmesnil commented on testsuite/integration/basic/src/test/java/org/wildfly/test/integration/microprofile/health/MicroProfileHealthTestBase.java:
```diff
@@ -60,15 +59,37 @@
+    @Deployment(name = EAR_ARCHIVE_NAME, managed = false)
+    public static Archive<?> deployEar() {
+        JavaArchive jarArchive = ShrinkWrap.create(JavaArchive.class, JAR_ARCHIVE_NAME)
+                .addClass(HealthConfigSource.class)
```
As mentioned in https://issues.redhat.com/browse/WFLY-12835?focusedCommentId=14115173&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14115173, the config source of a subdeployment will not support reading from another library in the EAR deployment

jmesnil commented on testsuite/integration/basic/src/test/java/org/wildfly/test/integration/microprofile/health/MicroProfileHealthTestBase.java:
```diff
@@ -84,40 +105,44 @@ public void testHealthCheckBeforeDeployment() throws Exception {
     @Test
     @InSequence(3)
-    public void testHealthCheckAfterUndeployment() throws Exception {
+    public void testHealthCheckBetweenDeployments() throws Exception {
```
Name of the method is not accurate, `testHealthCheckAfterUndeployment` is still valid

jmesnil commented on testsuite/integration/basic/src/test/resources/org/wildfly/test/integration/microprofile/config/smallrye/converter/jboss-deployment-structure.xml:
```diff
@@ -4,4 +4,4 @@
     </deployment>
-</jboss-deployment-structure>
\ No newline at end of file
+</jboss-deployment-structure>
```
this file does not need to be touched by this patch

jmesnil commented on testsuite/integration/basic/src/test/resources/org/wildfly/test/integration/microprofile/config/smallrye/management/config_source/jboss-deployment-structure.xml:
```diff
@@ -4,4 +4,4 @@
     </deployment>
-</jboss-deployment-structure>
\ No newline at end of file
+</jboss-deployment-structure>
```
this file does not need to be touched by this patch

jmesnil reviewed (changes requested):
Supporting MicroProfile Health Checks in EAR would still have restrictions related to MP ConfigSource.

The change inside WildFly code base looks fine (except the star imports), thanks.

Could you please update your PR to provide an additional test case to test EAR in addition to the existing one that test WAR?

cghislai committed:
Merge remote-tracking branch 'origin/master' into mp-ear

cghislai committed:
Cleanup & test ear deployment separately

cghislai commented:
@jmesnil thanks for your review. I addressed most of your comments, sorry for the formatting.
Unfortunately, after implementing the ear test, I noticed CDI.current() is not deterministic, and that prevents all healthchecks to be discovered. See https://issues.redhat.com/browse/WELD-2510 for reference.
If you know how I could perform a jndi lookup in this extension, please let me know. Im struggling with it.

hejulrik commented:
when and to which version of wildfly will this pull request be merged ? I would really be helped if this is fixed. 
 /Ulrik

cghislai commented:
@hejulrik sorry i kind of forgotten this one, and I doubt I will finish it. Im gonna close it.
I had trouble integrating with the wildfly deployment logic, and could not get a deterministic behaviour.

The workaround I use personally is to implement an application healthcheck endpoint with the same api. Then injecting the healtchecks using CDI I have no trouble with ear deployments.

hejulrik commented:
@cghislai ok 
in your  workaround, into what kind of class/bean/ejb are you injecting healtcheck-classes  ?

cghislai commented:
You can use any CDI bean I suppose. For instance an @ApplicationScoped
service or a @Stateless bean injects an Instance<HealthCheck>, iterates
over them, and returns the statusList to your jax-rs controller.

On Thu, Sep 24, 2020 at 3:41 PM hejulrik <notifications@github.com> wrote:

> @cghislai <https://github.com/cghislai> ok
> in your workaround, into what kind of class/bean/ejb are you injecting
> healtcheck-classes ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wildfly/wildfly/pull/13278#issuecomment-698351469>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADAPXZU7OAX4DYPPUTTCAUDSHNEATANCNFSM4M44Q2TA>
> .
>

hejulrik commented:
> You can use any CDI bean I suppose. For instance an @ApplicationScoped service or a @stateless bean injects an Instance<HealthCheck>, iterates over them, and returns the statusList to your jax-rs controller.
> […](#)
> On Thu, Sep 24, 2020 at 3:41 PM hejulrik ***@***.***> wrote: @cghislai <https://github.com/cghislai> ok in your workaround, into what kind of class/bean/ejb are you injecting healtcheck-classes ? — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#13278 (comment)](https://github.com/wildfly/wildfly/pull/13278#issuecomment-698351469)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADAPXZU7OAX4DYPPUTTCAUDSHNEATANCNFSM4M44Q2TA> .

ok thanks, so there is no magic to get the status reported to the global health status (for example  http://localhost:9990/health)
